### PR TITLE
Export special feature note to featureNames

### LINKED ISSUE #602: Feature names are not handled
Glyphs seems to (ab)use feature notes for setting features names (see _Names for Stylistic Sets_ section in https://glyphsapp.com/tutorials/stylistic-sets).

Currently glyphsLib will export any thing in the feature notes as a comment.

Test font: [Qahiri.zip](https://github.com/googlefonts/glyphsLib/files/4346480/Qahiri.zip) (`ss01` and `ss02` features).

### PR DESCRIPTION
**Pull request by jenskutilek**
The special feature note from Glyphs

```
Name: Single-storey a
```

will be exported to

```afdko
featureNames {
  name "Single-storey a";
};
```

Fixes #602

### CODE REVIEW

File: Lib/glyphsLib/builder/features.py [hunk 1]
```diff
@@ -89,6 +89,11 @@ def _to_ufo_features(font, ufo=None, generate_GDEF=False, skip_export_glyphs=Non
         if feature.notes:
             lines.append("# notes:")
             lines.extend("# " + line for line in feature.notes.splitlines())
+            feature_name = re.search("^Name: (.+)", feature.notes, flags=re.MULTILINE)
+            if feature_name and len(feature_name.groups()) == 1:
```
Comment (anthrotype) [inline Lib/glyphsLib/builder/features.py:93]: it's unlikely that you get more than one match, but in that case I'd simply take the last one

File: Lib/glyphsLib/builder/features.py [hunk 2]
```diff
@@ -89,6 +89,13 @@ def _to_ufo_features(font, ufo=None, generate_GDEF=False, skip_export_glyphs=Non
...
+                name = feature_name.groups()[0]
+                # Replace special chars backslash and doublequote for AFDKO syntax
+                name = name.replace("\x5c", r"\005c")
+                name = name.replace('"', r"\0022")
+                lines.extend(["featureNames {", '  name "%s";' % name, "};"])
```
Comment (anthrotype) [inline Lib/glyphsLib/builder/features.py:98]:
```suggestion
                lines.extend(["featureNames {", f'  name "{name}";', "};"])
```

File: Lib/glyphsLib/builder/features.py (hunk 1 above)
Comment (jenskutilek) [reply Lib/glyphsLib/builder/features.py]: Thanks! Changed in a5fb108.

File: Lib/glyphsLib/builder/features.py [hunk 3]
```diff
@@ -89,6 +89,13 @@ def _to_ufo_features(font, ufo=None, generate_GDEF=False, skip_export_glyphs=Non
         if feature.notes:
             lines.append("# notes:")
             lines.extend("# " + line for line in feature.notes.splitlines())
+            feature_name = re.search("^Name: (.+)", feature.notes, flags=re.MULTILINE)
+            if feature_name and feature_name.groups():
```
Comment (anthrotype) [inline Lib/glyphsLib/builder/features.py:93]: if the returned match object is not None, I think it must contain some non empty groups so the second test `feature_name.groups()` should be redundant

File: Lib/glyphsLib/builder/features.py [hunk 4]
```diff
@@ -89,6 +89,13 @@ def _to_ufo_features(font, ufo=None, generate_GDEF=False, skip_export_glyphs=Non
         if feature.notes:
             lines.append("# notes:")
             lines.extend("# " + line for line in feature.notes.splitlines())
+            feature_name = re.search("^Name: (.+)", feature.notes, flags=re.MULTILINE)
+            if feature_name:
+                name = feature_name.groups()[-1]
+                # Replace special chars backslash and doublequote for AFDKO syntax
+                name = name.replace("\x5c", r"\005c")
```
Comment (anthrotype) [inline Lib/glyphsLib/builder/features.py:96]: I think this would be more readable, I don't immediately remember that backslash codepoint is U+005C
```suggestion
                name = name.replace("\\", r"\005c")
```

**Comment by anthrotype** [comment]
the flake8 check failures appear to be unrelated to this PR
https://travis-ci.org/github/googlefonts/glyphsLib/jobs/699627388

@madig any idea where these are coming from?

File: Lib/glyphsLib/builder/features.py (hunk 4 above)
Comment (jenskutilek) [reply Lib/glyphsLib/builder/features.py]: Sure!

**Comment by jenskutilek** [comment]
> the flake8 check failures appear to be unrelated to this PR
> https://travis-ci.org/github/googlefonts/glyphsLib/jobs/699627388

Yes, they seem to appear on all recent pull requests ...

**Comment by madig** [comment]
Uh. No idea why these were missed before, except if flake8 actually never ran on those files. Uhm. Feel free to disable both errors in the config for this PR.

**Comment by madig** [comment]
Just updated master directly, you can rebase.

**Comment by jenskutilek** [comment]
Looking at the test font @khaledhosny posted on #602, this PR is not complete yet. It seems the long AFDKO feature name syntax can also be in the feature note. I thought it should go in the regular feature text.

**Comment by anthrotype** [comment]
the CLA bot is complaining with the cherry-pick of Nikolaus commit. @jenskutilek do you mind undoing that and instead do a regular merge (`git fetch origin && git merge origin/master`, provided that `origin` is the upstream repo).

File: setup.cfg [hunk 5]
```diff
@@ -61,7 +61,7 @@ formats = zip
 
 [flake8]
 select = B, C, E, F, W, T4, B9
-ignore = E203, E266, E501, W503
+ignore = E203, E266, E501, W503, E741
```
Comment (anthrotype) [inline setup.cfg:64]: hm, I would have rather fixed the offending code, rather than mute the flake8 warning globally

Comment (madig) [reply setup.cfg:64]: I've seen that warning every time you use a single-letter variable like in a list comprehension, so I decided to mute it for now.

File: tests/builder/builder_test.py [hunk 6]
```diff
@@ -1009,6 +1010,53 @@ def test_glyph_lib_Export_GDEF(self):
...
+            self.assertIn(r'name "Single\005cstorey \0022ä\0022";', ufo.features.text)
+
+    def test_glyph_lib_Export_feature_names_long_from_notes(self):
+        font = generate_minimal_font()
+        add_glyph(font, "a")
+        add_glyph(font, "a.ss01")
+        ss01 = GSFeature(name="ss01", code="sub a by a.ss01;")
+        font.features.append(ss01)
+        for note in (
+            'featureNames {\n  name 3 1 0x401 "Alternate {ä};";\n  name 3 1 0x409 "Alternate {};";\n};\n',
```
Comment (anthrotype) [inline tests/builder/builder_test.py:1037]: it would be good if you could wrap these very long strings here and further below into multiple lines

**Comment by anthrotype** [comment]
@madig ok to merge? Can you then tag a release with this after the merge? Thank you

**Comment by madig** [comment]
Idle thought: is it necessary to do anything when going UFO -> Glyphs?

**Comment by schriftgestalt** [comment]
If the "automatic" flag is round tripped, then it would be good to get the feature names out of the feature code and into the notes.

**Comment by justvanrossum** [comment]
Ping. Just running into this. Would be nice to get it merged.

**Comment by anthrotype** [comment]
I think the reason we haven't merged this yet is that it doesn't seem to fix https://github.com/googlefonts/glyphsLib/issues/603#issuecomment-646732838

I suppose we can merge this first and add a TODO for later..

File: Lib/glyphsLib/builder/features.py [hunk 7]
```diff
@@ -89,6 +89,22 @@ def _to_ufo_features(font, ufo=None, generate_GDEF=False, skip_export_glyphs=Non
...
             lines.extend("# " + line for line in feature.notes.splitlines())
+            feature_name = re.search(
+                "(featureNames {.+};)", feature.notes, flags=re.DOTALL
+            )
+            if feature_name:
+                name = feature_name.groups()[-1]
+                lines.extend(name.splitlines())
+            else:
+                feature_name = re.search(
+                    "^Name: (.+)", feature.notes, flags=re.MULTILINE
```
Comment (anthrotype) [inline Lib/glyphsLib/builder/features.py:100]: according to a comment from @schriftgestalt:

> only if the first line starts with "Name:", the rest of the line will be used

here you're matching any line in features.notes that starts with `Name:`

Comment (jenskutilek) [reply Lib/glyphsLib/builder/features.py]: Yep, that was before Georg clarified the behaviour :)

**Comment by jenskutilek** [comment]
> I think the reason we haven't merged this yet is that it doesn't seem to fix [#603 (comment)](https://github.com/googlefonts/glyphsLib/issues/603#issuecomment-646732838)
> 
> I suppose we can merge this first and add a TODO for later..

The roundtripping behaviour https://github.com/googlefonts/glyphsLib/pull/615#issuecomment-651056231 is not implemented yet. I'll see if I can do it today.

File: Lib/glyphsLib/builder/features.py [hunk 8]
```diff
@@ -630,6 +630,21 @@ def _process_feature_block(self):
...
+            feature_name = re.search(
+                r"(featureNames\s*{\s*name \"(.+)\";\s*};)",
+                feature.code,
+                flags=re.MULTILINE + re.DOTALL,
+            )
+            if feature_name:
+                statement, name = feature_name.groups()
+                if not re.search("^Name: (.+)", feature.notes):
+                    # Don't add if the name is already there
+                    feature.notes = f"Name: {name}\n" + feature.notes
```
Comment (jenskutilek) [inline Lib/glyphsLib/builder/features.py:645]: The feature names may go out of sync here, because we don't check if what's already in the feature notes contains the same name as the featureNames block. Do you think that's a problem?

File: Lib/glyphsLib/builder/features.py [hunk 9]
```diff
@@ -96,9 +96,7 @@ def _to_ufo_features(font, ufo=None, generate_GDEF=False, skip_export_glyphs=Non
                 name = feature_name.groups()[-1]
                 lines.extend(name.splitlines())
             else:
-                feature_name = re.search(
-                    "^Name: (.+)", feature.notes, flags=re.MULTILINE
-                )
+                feature_name = re.search("^Name: (.+)", feature.notes)
```
Comment (jenskutilek) [inline Lib/glyphsLib/builder/features.py:99]: Do you think the name should be removed from the notes comment after it has been written to the UFO feature code?

If it is left in the notes comment, we get code like this in the UFO:

```afdko
feature ss01 {
# notes:
# Name: Alternate g
featureNames {
  name "Alternate g";
};
# automatic
sub g by g.ss01;
} ss01;
```

Comment (anthrotype) [reply Lib/glyphsLib/builder/features.py:99]: Yes I think we should avoid the duplication

**Comment by jenskutilek** [comment]
> Why only for the automatic features?

Not sure ;) I changed it to process all features.

File: tests/builder/features_test.py [hunk 10]
```diff
@@ -147,6 +147,46 @@ def test_feature_name_automatic(tmpdir, ufo_module):
...
+    """
+    )
+
+    font, rtufo = roundtrip(ufo, tmpdir, ufo_module)
+
+    # Check code in Glyphs font
+    gs_feature = font.features[0]
+    assert gs_feature.automatic
+    assert gs_feature.code.strip() == "sub g by g.ss01;"
+    assert gs_feature.notes.strip() == "Name: Alternate g"
```
Comment (jenskutilek) [inline tests/builder/features_test.py:161]: This assertion fails:

```bash
        assert gs_feature.code.strip() == "sub g by g.ss01;"
>       assert gs_feature.notes.strip() == "Name: Alternate g"
E       AssertionError: assert '' == 'Name: Alternate g'
E         - Name: Alternate g

tests/builder/features_test.py:161: AssertionError
```

I have no idea why it doesn't work. When I convert a similar file with `ufo2glyphs`, everything looks fine. If you see some stupid error that I made, let me know ;)

File: Lib/glyphsLib/builder/features.py [hunk 11]
```diff
@@ -650,22 +650,22 @@ def _process_feature_block(self):
...
         )
         if feature_name:
             statement, name = feature_name.groups()
 
             # Don't add if the name is already there
-            if not re.search("^Name: (.+)", feature.notes, flags=re.DOTALL):
+            if not re.search(r"^Name: (.+)", feature.notes, flags=re.DOTALL):
                 feature.notes = "\n".join(
-                    [f"Name: {name}", feature.notes.lstrip()]
+                    [f"Name: {name}", notes_text]
```
Comment (jenskutilek) [inline Lib/glyphsLib/builder/features.py:663]: This is the code that sets the feature note, but sometimes doesn't ...?

File: Lib/glyphsLib/builder/features.py (hunk 9 above)
Comment (jenskutilek) [reply Lib/glyphsLib/builder/features.py]: Ok

**Comment by anthrotype** [comment]
I'm now leaning towards _not_ supporting the roundtripping of these special comments when going back from UFO to Glyphs.
If they are there in .glyphs source, we cut and paste them to UFO's features.fea code; when going back to .glyphs, we keep as is and don't do anything clever with the feature notes. I'm just trying to reduce the complexity here.

**Comment by anthrotype** [comment]
.. or maybe we only try to do that for "automatic" features as you did originally, to avoid surprises? sorry for the back and forth. I know how frustrating it can be when attempting to make _any_ changes in glyphsLib.

**Comment by schriftgestalt** [comment]
For automatic features, it would be good to properly round trip. 

And a small heads up how this is stored in Glyphs 3 files (indentation added for clarity):
```
{
	code = "sub a by b;";
	labels = (
		{
			language = ENG;
			value = "English Name";
		},
		{
			language = FOS;
			value = "Faroese Name";
		}
	);
	tag = ss01;
}
```

**Comment by jenskutilek** [comment]
I'm kind of lost here. I haven't looked at this in a while, but as I remember I can't get the roundtripping test to work because apparently at some point original data from the UFO lib is restored, overwriting any changes made at the UFO stage. It's hard to follow what exactly happens.

**Comment by khaledhosny** [comment]
With #603 fixed, can we split the round-tripping part into its own PR and keep this for the Glyphs -> UFO conversion and merge it?

**Comment by khaledhosny** [comment]
I have this branch rebased locally, should I force push here or open a new PR?

**Comment by anthrotype** [comment]
You can force push thanks

**Comment by khaledhosny** [comment]
> And a small heads up how this is stored in Glyphs 3 files (indentation added for clarity):

@schriftgestalt I’m trying to support the Glyphs 3 format, but I have a couple of questions:
1. How to map Glyphs’s label `language` to name table `langID`?
2. What to set `platformID` and `platEncID` to? (3, 1)?

**Comment by khaledhosny** [comment]
The regression tests failure is because feature names are now properly handled.

**Comment by khaledhosny** [comment]
This is ready, except for the mapping between glyphs language codes and name table language IDs (needed only for Glyphs 3 files).

**Comment by jenskutilek** [comment]
Thank you @khaledhosny for picking this up and finishing it!

**Comment by anthrotype** [comment]
Thanks Khaled!
